File: prototype/inheritance.js
// The base Employee class
module.exports = {};
function Employee() {
    this.name = "";
    this.dept = "general";
}
module.exports.Employee = Employee;
// Manager class inherits from Employee
function Manager() {
    Employee.call(this);
    this.reports = [];
}
Manager.prototype = Object.create(Employee.prototype);
Manager.prototype.constructor = Manager;
module.exports.Manager = Manager;

function WorkerBee() {
    Employee.call(this);
    this.projects = [];
}
WorkerBee.prototype = Object.create(Employee.prototype);
WorkerBee.prototype.constructor = WorkerBee;
module.exports.WorkerBee = WorkerBee;

function SalesPerson() {
    WorkerBee.call(this);
    this.quota = 100;
    this.dept = "Sales";
}
SalesPerson.prototype = Object.create(WorkerBee.prototype);
SalesPerson.prototype.constructor = SalesPerson;
module.exports.SalesPerson = SalesPerson;

function Engineer() {
    WorkerBee.call(this);
    this.machine = "";
    this.dept = "Engineering";
}
Engineer.prototype = Object.create(WorkerBee.prototype);
Engineer.prototype.constructor = Engineer;
module.exports.Engineer = Engineer;

File: prototype/Inheritance.java
import java.util.ArrayList;

public class Inheritance {
    public static void main(String[] args) {
        Manager mgr = new Manager();
        System.out.print(mgr.dept);
    }
}

class Employee {
    public String name = "";
    public String dept = "general";

    public Employee() {

    }
}

class Manager extends Employee {
    public ArrayList<Employee> reports = new ArrayList<>();

    public Manager() {

    }
}

class WorkerBee extends Employee {
    public ArrayList<String> projects = new ArrayList<>();

    public WorkerBee() {

    }
}

class SalesPerson extends WorkerBee {
    public Integer quota = 100;
    public String dept = "Sales";

    public SalesPerson() {

    }
}

class Engineer extends WorkerBee {
    public String machine = "";
    public String dept = "Engineering";

    public Engineer() {

    }

}